x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -942,22 +919,91 @@ public void onFailure(Exception e) {
+            public void onResponse(GetResponse response) {
+                if (response.isExists()) {
+                    logger.debug("found superseding token document [{}] in index [{}] by following the [{}] reference", response.getId(),
+                            response.getIndex(), refreshTokenStatus.getSupersedingDocId());
+                    final Tuple<UserToken, String> parsedTokens;
+                    try {
+                        parsedTokens = parseTokensFromDocument(response.getSource(), null);
+                    } catch (IllegalStateException | DateTimeException e) {
+                        logger.error("unable to decode existing user token", e);
+                        listener.onFailure(new ElasticsearchSecurityException("could not refresh the requested token", e));
+                        return;
+                    }
+                    listener.onResponse(parsedTokens);
+                } else {
+                    // We retry this since the creation of the superseding token document might already be in flight but not
+                    // yet completed, triggered by a refresh request that came a few milliseconds ago
+                    logger.info("could not find superseding token document from [{}] reference, retrying",
+                            refreshTokenStatus.getSupersedingDocId());
+                    maybeRetryOnFailure.accept(invalidGrantException("could not refresh the requested token"));
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                if (isShardNotAvailableException(e)) {
+                    logger.info("could not find superseding token document from reference [{}], retrying",
+                            refreshTokenStatus.getSupersedingDocId());
+                    maybeRetryOnFailure.accept(invalidGrantException("could not refresh the requested token"));
+                } else {
+                    logger.warn("could not find superseding token document from reference [{}]", refreshTokenStatus.getSupersedingDocId());
+                    onFailure.accept(invalidGrantException("could not refresh the requested token"));
+                }
+            }
+        });
+    }
+
+    private void getSupersedingTokenDocAsync(RefreshTokenStatus refreshTokenStatus, ActionListener<GetResponse> listener) {
+        final Optional<Tuple<Version, String>> versionAndSupersedingTokenDocId = tryUnpackVersionAndPayload(
+                refreshTokenStatus.getSupersedingDocId());
+        if (versionAndSupersedingTokenDocId.isPresent()) {
+            final Version supersedingTokenVersion = versionAndSupersedingTokenDocId.get().v1();
+            assert supersedingTokenVersion.onOrAfter(VERSION_TOKENS_INDEX_INTRODUCED);
+            final String supersedingTokenDocId = versionAndSupersedingTokenDocId.get().v2();
+            getTokenDocAsync(supersedingTokenDocId, securityTokensIndex, listener);
+        } else {
```

Comment:
This presumes that an old style tokens will never parse successfully into a Version & String. I don't know what guarantees we have that this is true.
They're both base64 encoded strings (with slightly different alphabets, the UUID uses the URL-safe encoder), so if the UUID never contains the binary values to produce `-` and `_`, then we will try and read the UUID bytes as a data input stream.
That will _usually_ fail, but I don't know that it is _impossible_ to have a random UUID that is also a valid data stream.

Length still seems safer to me. A UUID is always 16 bytes, so the base64 output will always be 24 bytes.
The new style tokens  must be longer than that because they have a version, and then the UUID written to a binary stream, and then base64 encoded.

Reply:
> This presumes that an old style tokens will never parse successfully into a Version & String

You're right! I was under the impression that random UUIDs were guaranteed to have a "-" in their name due to the bit shenanigans and a small sampling size, but this is not true.

> Length still seems safer to me.

Alright. I've adapted the implementation.